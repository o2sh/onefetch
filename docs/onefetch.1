.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.2.
.TH ONEFETCH "1" "septembre 2024" "onefetch 2.22.0" "User Commands"
.SH NAME
onefetch \- Command-line Git information tool
.SH SYNOPSIS
.B onefetch
[\fI\,OPTIONS\/\fR] [\fI\,INPUT\/\fR]
.SH DESCRIPTION
Command\-line Git information tool
.SS "Arguments:"
.IP
[INPUT]
.IP
Run as if onefetch was started in <input> instead of the current working directory
.SH OPTIONS
.HP
\fB\-h\fR, \fB\-\-help\fR
.IP
Print help (see a summary with '\-h')
.HP
\fB\-V\fR, \fB\-\-version\fR
.IP
Print version
.SS "INFO:"
.HP
\fB\-d\fR, \fB\-\-disabled\-fields\fR <FIELD>...
.IP
Allows you to disable FIELD(s) from appearing in the output
.HP
\fB\-\-no\-title\fR
.IP
Hides the title
.HP
\fB\-\-number\-of\-authors\fR <NUM>
.IP
Maximum NUM of authors to be shown
.IP
[default: 3]
.HP
\fB\-\-number\-of\-languages\fR <NUM>
.IP
Maximum NUM of languages to be shown
.IP
[default: 6]
.HP
\fB\-\-number\-of\-file\-churns\fR <NUM>
.IP
Maximum NUM of file churns to be shown
.IP
[default: 3]
.HP
\fB\-\-churn\-pool\-size\fR <NUM>
.IP
Minimum NUM of commits from HEAD used to compute the churn summary
.IP
By default, the actual value is non\-deterministic due to time\-based computation and will be displayed under the info title "Churn (NUM)"
.HP
\fB\-e\fR, \fB\-\-exclude\fR <EXCLUDE>...
.IP
Ignore all files & directories matching EXCLUDE
.HP
\fB\-\-no\-bots[=\fR<REGEX>]
.IP
Exclude [bot] commits. Use <REGEX> to override the default pattern
.HP
\fB\-\-no\-merges\fR
.IP
Ignores merge commits
.HP
\fB\-E\fR, \fB\-\-email\fR
.IP
Show the email address of each author
.HP
\fB\-\-http\-url\fR
.IP
Display repository URL as HTTP
.HP
\fB\-\-hide\-token\fR
.IP
Hide token in repository URL
.HP
\fB\-\-include\-hidden\fR
.IP
Count hidden files and directories
.HP
\fB\-T\fR, \fB\-\-type\fR <TYPE>...
.IP
Filters output by language type
.IP
[default: programming markup]
[possible values: programming, markup, prose, data]
.SS "TEXT FORMATTING:"
.HP
\fB\-t\fR, \fB\-\-text\-colors\fR <X>...
.IP
Changes the text colors (X X X...)
.IP
Goes in order of title, ~, underline, subtitle, colon, and info
.IP
For example:
.IP
\&'\-\-text\-colors 9 10 11 12 13 14'
.HP
\fB\-z\fR, \fB\-\-iso\-time\fR
.IP
Use ISO 8601 formatted timestamps
.HP
\fB\-\-number\-separator\fR <SEPARATOR>
.IP
Which thousands SEPARATOR to use
.IP
[default: plain]
[possible values: plain, comma, space, underscore]
.HP
\fB\-\-no\-bold\fR
.IP
Turns off bold formatting
.SS "ASCII:"
.HP
\fB\-\-ascii\-input\fR <STRING>
.IP
Takes a non\-empty STRING as input to replace the ASCII logo
.IP
It is possible to pass a generated STRING by command substitution
.IP
For example:
.IP
\&'\-\-ascii\-input "$(fortune | cowsay \fB\-W\fR 25)"'
.HP
\fB\-c\fR, \fB\-\-ascii\-colors\fR <X>...
.IP
Colors (X X X...) to print the ascii art
.HP
\fB\-a\fR, \fB\-\-ascii\-language\fR <LANGUAGE>
.IP
Which LANGUAGE's ascii art to print
.HP
\fB\-\-true\-color\fR <WHEN>
.IP
Specify when to use true color
.IP
If set to auto: true color will be enabled if supported by the terminal
.IP
[default: auto]
[possible values: auto, never, always]
.SS "IMAGE:"
.HP
\fB\-i\fR, \fB\-\-image\fR <IMAGE>
.IP
Path to the IMAGE file
.HP
\fB\-\-image\-protocol\fR <PROTOCOL>
.IP
Which image PROTOCOL to use
.IP
[possible values: kitty, sixel, iterm]
.HP
\fB\-\-color\-resolution\fR <VALUE>
.IP
VALUE of color resolution to use with SIXEL backend
.IP
[default: 16]
[possible values: 16, 32, 64, 128, 256]
.SS "VISUALS:"
.HP
\fB\-\-no\-color\-palette\fR
.IP
Hides the color palette
.HP
\fB\-\-no\-art\fR
.IP
Hides the ascii art or image if provided
.HP
\fB\-\-nerd\-fonts\fR
.IP
Use Nerd Font icons
.IP
Replaces language chips with Nerd Font icons
.SS "DEVELOPER:"
.HP
\fB\-o\fR, \fB\-\-output\fR <FORMAT>
.IP
Outputs Onefetch in a specific format
.IP
[possible values: json, yaml]
.HP
\fB\-\-generate\fR <SHELL>
.IP
If provided, outputs the completion file for given SHELL
.IP
[possible values: bash, elvish, fish, powershell, zsh]
.SS "OTHER:"
.HP
\fB\-l\fR, \fB\-\-languages\fR
.IP
Prints out supported languages
.HP
\fB\-p\fR, \fB\-\-package\-managers\fR
.IP
Prints out supported package managers
