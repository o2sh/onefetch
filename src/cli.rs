use crate::info::langs::language::{Language, LanguageType};
use crate::info::utils::info_field::InfoType;
use crate::ui::printer::SerializationFormat;
use anyhow::Result;
use clap::builder::PossibleValuesParser;
use clap::builder::TypedValueParser as _;
use clap::{value_parser, Args, Command, Parser, ValueHint};
use clap_complete::{generate, Generator, Shell};
use num_format::CustomFormat;
use onefetch_image::ImageProtocol;
use onefetch_manifest::ManifestType;
use regex::Regex;
use serde::Serialize;
use std::env;
use std::io;
use std::path::PathBuf;
use std::str::FromStr;
use strum::IntoEnumIterator;

const COLOR_RESOLUTIONS: [&str; 5] = ["16", "32", "64", "128", "256"];
pub const NO_BOTS_DEFAULT_REGEX_PATTERN: &str = r"(?:-|\s)[Bb]ot$|\[[Bb]ot\]";

#[derive(Clone, Debug, Parser, PartialEq, Eq)]
#[command(version, about)]
pub struct CliOptions {
    /// Run as if onefetch was started in <input> instead of the current working directory
    #[arg(default_value = ".", hide_default_value = true, value_hint = ValueHint::DirPath)]
    pub input: PathBuf,
    #[command(flatten)]
    pub info: InfoCliOptions,
    #[command(flatten)]
    pub text_formatting: TextForamttingCliOptions,
    #[command(flatten)]
    pub ascii: AsciiCliOptions,
    #[command(flatten)]
    pub image: ImageCliOptions,
    #[command(flatten)]
    pub visuals: VisualsCliOptions,
    #[command(flatten)]
    pub developer: DeveloperCliOptions,
    #[command(flatten)]
    pub other: OtherCliOptions,
}

#[derive(Clone, Debug, Args, PartialEq, Eq)]
#[command(next_help_heading = "INFO")]
pub struct InfoCliOptions {
    /// Allows you to disable FIELD(s) from appearing in the output
    #[arg(
        long,
        short,
        num_args = 1..,
        hide_possible_values = true,
        value_enum,
        value_name = "FIELD"
    )]
    pub disabled_fields: Vec<InfoType>,
    /// Hides the title
    #[arg(long)]
    pub no_title: bool,
    /// Maximum NUM of authors to be shown
    #[arg(long, default_value_t = 3usize, value_name = "NUM")]
    pub number_of_authors: usize,
    /// Maximum NUM of languages to be shown
    #[arg(long, default_value_t = 6usize, value_name = "NUM")]
    pub number_of_languages: usize,
    /// Maximum NUM of file churns to be shown
    #[arg(long, default_value_t = 3usize, value_name = "NUM")]
    pub number_of_file_churns: usize,
    /// Minimum NUM of commits from HEAD used to compute the churn summary
    ///
    /// By default, the actual value is non-deterministic due to time-based computation
    /// and will be displayed under the info title "Churn (NUM)"
    #[arg(long, value_name = "NUM")]
    pub churn_pool_size: Option<usize>,
    /// Ignore all files & directories matching EXCLUDE
    #[arg(long, short, num_args = 1..)]
    pub exclude: Vec<String>,
    /// Exclude [bot] commits. Use <REGEX> to override the default pattern
    #[arg(
        long,
        num_args = 0..=1,
        require_equals = true,
        default_missing_value = NO_BOTS_DEFAULT_REGEX_PATTERN,
        value_name = "REGEX"
    )]
    pub no_bots: Option<MyRegex>,
    /// Ignores merge commits
    #[arg(long)]
    pub no_merges: bool,
    /// Show the email address of each author
    #[arg(long, short = 'E')]
    pub email: bool,
    /// Display repository URL as HTTP
    #[arg(long)]
    pub http_url: bool,
    /// Hide token in repository URL
    #[arg(long)]
    pub hide_token: bool,
    /// Count hidden files and directories
    #[arg(long)]
    pub include_hidden: bool,
    /// Filters output by language type
    #[arg(
        long,
        num_args = 1..,
        default_values = &["programming", "markup"],
        short = 'T',
        value_enum,
    )]
    pub r#type: Vec<LanguageType>,
}

#[derive(Clone, Debug, Args, PartialEq, Eq)]
#[command(next_help_heading = "ASCII")]
pub struct AsciiCliOptions {
    /// Takes a non-empty STRING as input to replace the ASCII logo
    ///
    /// It is possible to pass a generated STRING by command substitution
    ///
    /// For example:
    ///
    /// '--ascii-input "$(fortune | cowsay -W 25)"'
    #[arg(long, value_name = "STRING", value_hint = ValueHint::CommandString)]
    pub ascii_input: Option<String>,
    /// Colors (X X X...) to print the ascii art
    #[arg(
        long,
        num_args = 1..,
        value_name = "X",
        short = 'c',
        value_parser = value_parser!(u8).range(..16),
    )]
    pub ascii_colors: Vec<u8>,
    /// Which LANGUAGE's ascii art to print
    #[arg(
        long,
        short,
        value_name = "LANGUAGE",
        value_enum,
        hide_possible_values = true
    )]
    pub ascii_language: Option<Language>,
    /// Specify when to use true color
    ///
    /// If set to auto: true color will be enabled if supported by the terminal
    #[arg(long, default_value = "auto", value_name = "WHEN", value_enum)]
    pub true_color: When,
}

#[derive(Clone, Debug, Args, PartialEq, Eq)]
#[command(next_help_heading = "IMAGE")]
pub struct ImageCliOptions {
    /// Path to the IMAGE file
    #[arg(long, short, value_hint = ValueHint::FilePath)]
    pub image: Option<PathBuf>,
    /// Which image PROTOCOL to use
    #[arg(long, value_enum, requires = "image", value_name = "PROTOCOL")]
    pub image_protocol: Option<ImageProtocol>,
    /// VALUE of color resolution to use with SIXEL backend
    #[arg(
        long,
        value_name = "VALUE",
        requires = "image",
        default_value_t = 16usize,
        value_parser = PossibleValuesParser::new(COLOR_RESOLUTIONS)
            .map(|s| s.parse::<usize>().unwrap())
    )]
    pub color_resolution: usize,
}

#[derive(Clone, Debug, Args, PartialEq, Eq)]
#[command(next_help_heading = "TEXT FORMATTING")]
pub struct TextForamttingCliOptions {
    /// Changes the text colors (X X X...)
    ///
    /// Goes in order of title, ~, underline, subtitle, colon, and info
    ///
    /// For example:
    ///
    /// '--text-colors 9 10 11 12 13 14'
    #[arg(
        long,
        short,
        value_name = "X",
        value_parser = value_parser!(u8).range(..16),
        num_args = 1..=6
    )]
    pub text_colors: Vec<u8>,
    /// Use ISO 8601 formatted timestamps
    #[arg(long, short = 'z')]
    pub iso_time: bool,
    /// Which thousands SEPARATOR to use
    #[arg(long, value_name = "SEPARATOR", default_value = "plain", value_enum)]
    pub number_separator: NumberSeparator,
    /// Turns off bold formatting
    #[arg(long)]
    pub no_bold: bool,
}
#[derive(Clone, Debug, Args, PartialEq, Eq, Default)]
#[command(next_help_heading = "VISUALS")]
pub struct VisualsCliOptions {
    /// Hides the color palette
    #[arg(long)]
    pub no_color_palette: bool,
    /// Hides the ascii art or image if provided
    #[arg(long)]
    pub no_art: bool,
    /// Use Nerd Font icons
    ///
    /// Replaces language chips with Nerd Font icons
    #[arg(long)]
    pub nerd_fonts: bool,
}

#[derive(Clone, Debug, Args, PartialEq, Eq, Default)]
#[command(next_help_heading = "DEVELOPER")]
pub struct DeveloperCliOptions {
    /// Outputs Onefetch in a specific format
    #[arg(long, short, value_name = "FORMAT", value_enum)]
    pub output: Option<SerializationFormat>,
    /// If provided, outputs the completion file for given SHELL
    #[arg(long = "generate", value_name = "SHELL", value_enum)]
    pub completion: Option<Shell>,
}

#[derive(Clone, Debug, Args, PartialEq, Eq, Default)]
#[command(next_help_heading = "OTHER")]
pub struct OtherCliOptions {
    /// Prints out supported languages
    #[arg(long, short)]
    pub languages: bool,
    /// Prints out supported package managers
    #[arg(long, short)]
    pub package_managers: bool,
}

impl Default for CliOptions {
    fn default() -> CliOptions {
        CliOptions {
            input: PathBuf::from("."),
            info: InfoCliOptions::default(),
            text_formatting: TextForamttingCliOptions::default(),
            visuals: VisualsCliOptions::default(),
            ascii: AsciiCliOptions::default(),
            image: ImageCliOptions::default(),
            developer: DeveloperCliOptions::default(),
            other: OtherCliOptions::default(),
        }
    }
}

impl Default for InfoCliOptions {
    fn default() -> Self {
        InfoCliOptions {
            number_of_authors: 3,
            number_of_languages: 6,
            number_of_file_churns: 3,
            churn_pool_size: Option::default(),
            exclude: Vec::default(),
            no_bots: Option::default(),
            no_merges: Default::default(),
            email: Default::default(),
            http_url: Default::default(),
            hide_token: Default::default(),
            include_hidden: Default::default(),
            r#type: vec![LanguageType::Programming, LanguageType::Markup],
            disabled_fields: Vec::default(),
            no_title: Default::default(),
        }
    }
}

impl Default for TextForamttingCliOptions {
    fn default() -> Self {
        TextForamttingCliOptions {
            text_colors: Default::default(),
            iso_time: Default::default(),
            number_separator: NumberSeparator::Plain,
            no_bold: Default::default(),
        }
    }
}

impl Default for AsciiCliOptions {
    fn default() -> Self {
        AsciiCliOptions {
            ascii_input: Option::default(),
            ascii_colors: Vec::default(),
            ascii_language: Option::default(),
            true_color: When::Auto,
        }
    }
}
impl Default for ImageCliOptions {
    fn default() -> Self {
        ImageCliOptions {
            image: Option::default(),
            image_protocol: Default::default(),
            color_resolution: 16,
        }
    }
}

pub fn print_supported_languages() -> Result<()> {
    for l in Language::iter() {
        println!("{l}");
    }

    Ok(())
}

pub fn print_supported_package_managers() -> Result<()> {
    for p in ManifestType::iter() {
        println!("{p}");
    }

    Ok(())
}

pub fn is_truecolor_terminal() -> bool {
    env::var("COLORTERM")
        .map(|colorterm| colorterm == "truecolor" || colorterm == "24bit")
        .unwrap_or(false)
}

pub fn get_git_version() -> String {
    let version = std::process::Command::new("git").arg("--version").output();

    match version {
        Ok(v) => String::from_utf8_lossy(&v.stdout).replace('\n', ""),
        Err(_) => String::new(),
    }
}

pub fn print_completions<G: Generator>(gen: G, cmd: &mut Command) {
    generate(gen, cmd, cmd.get_name().to_string(), &mut io::stdout());
}

#[derive(clap::ValueEnum, Clone, PartialEq, Eq, Debug)]
pub enum When {
    Auto,
    Never,
    Always,
}

#[derive(clap::ValueEnum, Clone, PartialEq, Eq, Debug, Serialize, Copy)]
pub enum NumberSeparator {
    Plain,
    Comma,
    Space,
    Underscore,
}

impl NumberSeparator {
    fn separator(&self) -> &'static str {
        match self {
            Self::Plain => "",
            Self::Comma => ",",
            Self::Space => "\u{202f}",
            Self::Underscore => "_",
        }
    }

    pub fn get_format(&self) -> CustomFormat {
        num_format::CustomFormat::builder()
            .grouping(num_format::Grouping::Standard)
            .separator(self.separator())
            .build()
            .unwrap()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_default_config() {
        let config: CliOptions = CliOptions::default();
        assert_eq!(config, CliOptions::parse_from(["onefetch"]));
    }

    #[test]
    fn test_custom_config() {
        let config: CliOptions = CliOptions {
            input: PathBuf::from("/tmp/folder"),
            info: InfoCliOptions {
                number_of_authors: 4,
                no_merges: true,
                disabled_fields: vec![InfoType::Version, InfoType::URL],
                ..Default::default()
            },
            ascii: AsciiCliOptions {
                ascii_colors: vec![5, 0],
                ascii_language: Some(Language::Lisp),
                ..Default::default()
            },
            visuals: VisualsCliOptions {
                no_art: true,
                ..Default::default()
            },
            ..Default::default()
        };

        assert_eq!(
            config,
            CliOptions::parse_from([
                "onefetch",
                "/tmp/folder",
                "--number-of-authors",
                "4",
                "--no-merges",
                "--ascii-colors",
                "5",
                "0",
                "--disabled-fields",
                "version",
                "url",
                "--no-art",
                "--ascii-language",
                "lisp"
            ])
        );
    }

    #[test]
    fn test_config_with_image_protocol_but_no_image() {
        assert!(CliOptions::try_parse_from(["onefetch", "--image-protocol", "sixel"]).is_err())
    }

    #[test]
    fn test_config_with_color_resolution_but_no_image() {
        assert!(CliOptions::try_parse_from(["onefetch", "--color-resolution", "32"]).is_err())
    }

    #[test]
    fn test_config_with_ascii_colors_but_out_of_bounds() {
        assert!(CliOptions::try_parse_from(["onefetch", "--ascii-colors", "17"]).is_err())
    }

    #[test]
    fn test_config_with_text_colors_but_out_of_bounds() {
        assert!(CliOptions::try_parse_from(["onefetch", "--text-colors", "17"]).is_err())
    }
}

#[derive(Clone, Debug)]
pub struct MyRegex(pub Regex);

impl Eq for MyRegex {}

impl PartialEq for MyRegex {
    fn eq(&self, other: &MyRegex) -> bool {
        self.0.as_str() == other.0.as_str()
    }
}

impl FromStr for MyRegex {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        Ok(MyRegex(Regex::new(s)?))
    }
}
