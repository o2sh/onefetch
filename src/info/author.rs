use crate::{
    cli::NumberSeparator,
    info::{format_number, utils::git::CommitMetrics, utils::info_field::InfoField},
};
use owo_colors::{DynColors, OwoColorize};
use serde::Serialize;
use std::fmt::Write;

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Author {
    pub name: String,
    email: Option<String>,
    nbr_of_commits: usize,
    contribution: usize,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl Author {
    pub fn new(
        name: String,
        email: Option<String>,
        nbr_of_commits: usize,
        total_nbr_of_commits: usize,
        number_separator: NumberSeparator,
    ) -> Self {
        let contribution =
            (nbr_of_commits as f32 * 100. / total_nbr_of_commits as f32).round() as usize;
        Self {
            name,
            email,
            nbr_of_commits,
            contribution,
            number_separator,
        }
    }
}

impl std::fmt::Display for Author {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        if let Some(email) = &self.email {
            write!(
                f,
                "{}% {} <{}> {}",
                self.contribution,
                self.name,
                email,
                format_number(&self.nbr_of_commits, self.number_separator)
            )
        } else {
            write!(
                f,
                "{}% {} {}",
                self.contribution,
                self.name,
                format_number(&self.nbr_of_commits, self.number_separator)
            )
        }
    }
}

#[derive(Serialize)]
pub struct AuthorsInfo {
    pub authors: Vec<Author>,
    #[serde(skip_serializing)]
    pub info_color: DynColors,
}

impl AuthorsInfo {
    pub fn new(info_color: DynColors, commit_metrics: &CommitMetrics) -> Self {
        let authors = commit_metrics.authors_to_display.clone();
        Self {
            authors,
            info_color,
        }
    }
}

impl std::fmt::Display for AuthorsInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let mut authors_info = String::new();

        let pad = self.title().len() + 2;

        for (i, author) in self.authors.iter().enumerate() {
            let author_str = author.color(self.info_color);

            if i == 0 {
                write!(authors_info, "{author_str}")?;
            } else {
                write!(authors_info, "\n{:<width$}{}", "", author_str, width = pad)?;
            }
        }

        write!(f, "{authors_info}")
    }
}

#[typetag::serialize]
impl InfoField for AuthorsInfo {
    fn value(&self) -> String {
        self.to_string()
    }

    fn title(&self) -> String {
        let mut title: String = "Author".into();
        if self.authors.len() > 1 {
            title.push('s')
        }
        title
    }

    fn should_color(&self) -> bool {
        false
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_display_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        assert_eq!(author.to_string(), "75% John Doe <john.doe@email.com> 1500");
    }

    #[test]
    fn test_display_author_with_no_email() {
        let author = Author::new("John Doe".into(), None, 1500, 2000, NumberSeparator::Plain);

        assert_eq!(author.to_string(), "75% John Doe 1500");
    }

    #[test]
    fn test_authors_info_title_with_one_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            info_color: DynColors::Rgb(255, 0, 0),
            authors: vec![author],
        };

        assert_eq!(authors_info.title(), "Author");
    }

    #[test]
    fn test_authors_info_title_with_two_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            info_color: DynColors::Rgb(255, 0, 0),
            authors: vec![author, author_2],
        };

        assert_eq!(authors_info.title(), "Authors");
    }

    #[test]
    fn test_author_info_value_with_one_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            info_color: DynColors::Rgb(255, 0, 0),
            authors: vec![author],
        };

        assert_eq!(
            authors_info.value(),
            "75% John Doe <john.doe@email.com> 1500"
                .color(DynColors::Rgb(255, 0, 0))
                .to_string()
        );
    }

    #[test]
    fn test_author_info_value_with_two_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            info_color: DynColors::Rgb(255, 0, 0),
            authors: vec![author, author_2],
        };

        assert!(authors_info.value().contains(
            &"75% John Doe <john.doe@email.com> 1500"
                .color(DynColors::Rgb(255, 0, 0))
                .to_string()
        ));

        assert!(authors_info.value().contains(
            &"80% Roberto Berto 240"
                .color(DynColors::Rgb(255, 0, 0))
                .to_string()
        ));
    }
}
