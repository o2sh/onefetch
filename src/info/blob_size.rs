use super::utils::format_number;
use crate::info::langs::get_total_size;
use crate::info::langs::language::Language;
use crate::{cli::NumberSeparator, info::utils::info_field::InfoField};
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlobSizeInfo {
    pub size: usize,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl BlobSizeInfo {
    pub fn new(size_by_language: &[(Language, usize)], number_separator: NumberSeparator) -> Self {
        let size = get_total_size(size_by_language);
        Self {
            size,
            number_separator,
        }
    }
}

#[typetag::serialize]
impl InfoField for BlobSizeInfo {
    fn value(&self) -> String {
        format_number(&self.size, self.number_separator)
    }

    fn title(&self) -> String {
        "Total blob size".into()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_display_loc_info() {
        let loc_info = BlobSizeInfo {
            size: 1235,
            number_separator: NumberSeparator::Plain,
        };

        assert_eq!(loc_info.value(), "1235".to_string());
    }
}
