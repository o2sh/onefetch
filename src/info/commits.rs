use serde::Serialize;

use crate::{
    cli::NumberSeparator,
    info::{format_number, utils::git::CommitMetrics, utils::info_field::InfoField},
};

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CommitsInfo {
    pub number_of_commits: usize,
    is_shallow: bool,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl CommitsInfo {
    pub fn new(commit_metrics: &CommitMetrics, number_separator: NumberSeparator) -> Self {
        Self {
            number_of_commits: commit_metrics.total_number_of_commits,
            is_shallow: commit_metrics.is_shallow,
            number_separator,
        }
    }
}

#[typetag::serialize]
impl InfoField for CommitsInfo {
    fn value(&self) -> String {
        format!(
            "{}{}",
            format_number(&self.number_of_commits, self.number_separator),
            self.is_shallow.then_some(" (shallow)").unwrap_or_default()
        )
    }

    fn title(&self) -> String {
        "Commits".into()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_display_commits_info() {
        let commits_info = CommitsInfo {
            number_of_commits: 3,
            is_shallow: false,
            number_separator: NumberSeparator::Plain,
        };

        assert_eq!(commits_info.value(), "3".to_string());
    }

    #[test]
    fn test_display_commits_info_shallow() {
        let commits_info = CommitsInfo {
            number_of_commits: 2,
            is_shallow: true,
            number_separator: NumberSeparator::Plain,
        };

        assert_eq!(commits_info.value(), "2 (shallow)".to_string());
    }
}
