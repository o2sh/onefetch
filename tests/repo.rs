use anyhow::Result;
use gix::{open, Repository, ThreadSafeRepository};
use onefetch::cli::{CliOptions, InfoCliOptions, TextForamttingCliOptions};
use onefetch::info::{build_info, get_work_dir};

fn repo(name: &str) -> Result<Repository> {
    let repo_path = gix_testtools::scripted_fixture_read_only(name).unwrap();
    let safe_repo = ThreadSafeRepository::open_opts(repo_path, open::Options::isolated())?;
    Ok(safe_repo.to_thread_local())
}

pub fn named_repo(fixture: &str, name: &str) -> Result<Repository> {
    let repo_path = gix_testtools::scripted_fixture_read_only(fixture)
        .unwrap()
        .join(name);
    let safe_repo = ThreadSafeRepository::open_opts(repo_path, open::Options::isolated())?;
    Ok(safe_repo.to_thread_local())
}

#[test]
fn test_bare_repo() -> Result<()> {
    let repo = repo("make_bare_repo.sh")?;
    let work_dir = get_work_dir(&repo);
    assert!(
        work_dir.is_err(),
        "oops, info was returned on a bare git repo"
    );
    assert_eq!(
        work_dir.unwrap_err().to_string(),
        "please run onefetch inside of a non-bare git repository"
    );
    Ok(())
}

#[test]
fn test_repo() -> Result<()> {
    let repo = repo("make_repo.sh")?;
    let config: CliOptions = CliOptions {
        input: repo.path().to_path_buf(),
        info: InfoCliOptions {
            email: true,
            churn_pool_size: Some(10),
            ..Default::default()
        },
        text_formatting: TextForamttingCliOptions {
            iso_time: true,
            ..Default::default()
        },
        ..Default::default()
    };
    let info = build_info(&config)?;
    insta::assert_json_snapshot!(
        info,
        {
            ".title.gitVersion" => "git version",
            ".infoFields[].HeadInfo.headRefs.shortCommitId" => "short commit",
        }
    );

    Ok(())
}

#[test]
fn test_repo_without_remote() -> Result<()> {
    let repo = repo("make_repo_without_remote.sh")?;
    let config: CliOptions = CliOptions {
        input: repo.path().to_path_buf(),
        ..Default::default()
    };
    let info = build_info(&config);
    assert!(info.is_ok());

    Ok(())
}

#[test]
fn test_partial_repo() -> Result<()> {
    let repo = named_repo("make_partial_repo.sh", "partial")?;
    let config: CliOptions = CliOptions {
        input: repo.path().to_path_buf(),
        ..Default::default()
    };
    let _info = build_info(&config).expect("no error");
    Ok(())
}

#[test]
fn test_repo_with_pre_epoch_dates() -> Result<()> {
    let repo = repo("make_pre_epoch_repo.sh")?;
    let config: CliOptions = CliOptions {
        input: repo.path().to_path_buf(),
        ..Default::default()
    };
    let _info = build_info(&config).expect("no error");
    Ok(())
}
