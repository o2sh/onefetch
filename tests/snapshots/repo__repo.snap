---
source: tests/repo.rs
expression: info
---
{
  "title": {
    "gitUsername": "onefetch-committer-name",
    "gitVersion": "git version"
  },
  "infoFields": [
    {
      "ProjectInfo": {
        "repoName": "repo",
        "numberOfBranches": 0,
        "numberOfTags": 1
      }
    },
    {
      "DescriptionInfo": {
        "description": "Amazing tool"
      }
    },
    {
      "HeadInfo": {
        "headRefs": {
          "shortCommitId": "short commit",
          "refs": [
            "main"
          ]
        }
      }
    },
    {
      "PendingInfo": {
        "added": 2,
        "deleted": 0,
        "modified": 1
      }
    },
    {
      "VersionInfo": {
        "version": "tag1"
      }
    },
    {
      "CreatedInfo": {
        "creationDate": "2000-01-02T00:00:00Z"
      }
    },
    {
      "LanguagesInfo": {
        "languagesWithPercentage": [
          {
            "language": "Rust",
            "percentage": 100.0
          }
        ]
      }
    },
    {
      "DependenciesInfo": {
        "dependencies": "1 (Cargo)"
      }
    },
    {
      "AuthorsInfo": {
        "authors": [
          {
            "name": "Author Four",
            "email": "author4@example.org",
            "nbrOfCommits": 1,
            "contribution": 25
          },
          {
            "name": "Author One",
            "email": "author1@example.org",
            "nbrOfCommits": 1,
            "contribution": 25
          },
          {
            "name": "Author Three",
            "email": "author3@example.org",
            "nbrOfCommits": 1,
            "contribution": 25
          }
        ]
      }
    },
    {
      "LastChangeInfo": {
        "lastChange": "2000-01-02T00:00:00Z"
      }
    },
    {
      "ContributorsInfo": {
        "totalNumberOfAuthors": 4
      }
    },
    {
      "UrlInfo": {
        "repoUrl": "https://github.com/user/repo.git"
      }
    },
    {
      "CommitsInfo": {
        "numberOfCommits": 4,
        "isShallow": false
      }
    },
    {
      "ChurnInfo": {
        "file_churns": [
          {
            "filePath": "code.rs",
            "nbrOfCommits": 4
          }
        ],
        "churn_pool_size": 4
      }
    },
    {
      "LocInfo": {
        "linesOfCode": 4
      }
    },
    {
      "SizeInfo": {
        "repoSize": "22 B",
        "fileCount": 1
      }
    },
    {
      "LicenseInfo": {
        "license": "MIT"
      }
    }
  ]
}
